### ISSUE TITLE
How to revoke certificate?

### ISSUE THREAD

**Issue by hmatt843** [ebekker/ACMESharp#103]
I'm using Windows Server 2012 R2 and IIS. I used ACMESharp to get a LetsEncrypt certificate for my site, and it worked great, but now I need to revoke it. Is the only way to revoke it really to install Linux on the machine? What are my options? Please help.

**Comment by bseddon** [comment]
I've just looked at the ACMESharp code and there does not appear to be a function to access the Lets Encrypt certificate revocation mechanism at the moment.  It looks like this function will need to be added to the ACMESharp client class, exposed as a PowerShell command and unit tests added. In the meantime it does look like your options to revoke a certificate are limited and probably involve using the Linux client.

**Comment by hmatt843** [comment]
Based on screenshots I took from when I was signing up, http://i.imgur.com/UqQ0VWn.png It looks like there are 2 expiration dates? One of which has already passed in February 22 2016 (I signed up on February 15 2016), and one which is going to pass on December 2016. 

Do you know how this impacts my case? I am new to certificate stuff so I feel like I am reading everything in Greek!

**Comment by bseddon** [comment]
Normally certificate generation is not so complicated.  If you use a paid-for service from your domain name provider you will generate a certificate signing request then paste it into the vendors page and receive your certificate.  The domain name provider doesn't need to validate your domain because they know it.

Using Let's Encrypt is more complicated because: a) they want to be able to handle domain validation; b) want to allow multiple domains to be validated by a single certificate; and c) have certificates with short expiry terms.  So when using Let's Encrypt you get to see more of the inner workings. Perhaps that the payment in kind.

I see the same behaviour you are seeing.  If I issue a Get-ACMEIdentifier command *after* the certificate has been generated but before the identifier has been updated, then likke you I see: a) it's showing a pending state; and b) the date is the date the identifier was generated.

When I then issue an Update-ACMEIdentifier command for the same alias (after the certificate has been generated) the information shows a date in December.  The date I created an alias was Feb 28th and the expiry date now shows Dec 17th.  Your date was Feb 22nd and is now Dec 11th.  The number of days is the same (294).  I see this as the date until which this identifier can be used to create a certificate.  Let's Encrypt certificates expire after 90 days.  Maybe its some longer date so certificates are not expired early.

I do see a difference in our outputs: I see a valid uri after updating but there is no uri in your screenshot.

**Comment by hmatt843** [comment]
Thank you. Do you happen to have any advice on signing up for the www and non-www version of your URL using ACMESharp? The reason why I even wanted to revoke my certificate is because https://mysite.com works, but https://www.mysite.com , and I think that's because of the commands I typed when signing up.

I typed: 
`New-ACMEIdentifier -Dns mysite.com -Alias mysite.com_Alias`

And what I'm guessing I should have typed was something like: 

`New-ACMEIdentifier -Dns mysite.com www.mysite.com -Alias mysite.com_Alias www.mysite.com_Alias`

...but that is giving me syntax errors. I tried a bunch of different variations of this, but nothing is working syntax-wise. I am starting to doubt that it is even possible to do this with ACMESharp? And that makes me wonder what I am supposed to do to get BOTH the www and non-www version of my site working. Do you think I have to have a separate certificate for each version of the URL? That seems weird to me. What might you suggest?

**Comment by bseddon** [comment]
Yes, you are able to generate certificates that support multiple domains like the example you give.  Look at the 'Subject Alternative Names' section on the [Quick start](https://github.com/ebekker/ACMESharp/wiki/Quick-Start) page for the syntax to use.  Also see this thread which reviews the syntax issues that arise when you have only on alternative name.

The trick is that each domain to be included in a certificate must be validated separately.  That is, each will have its own valid identifier.  You are not adding a list of domains to to the AlternativeIdentifierRefs parameter, you are adding a list of pre-validated aliases.

**Comment by hmatt843** [comment]
To be clear, I would need to do this part:

`PS> New-ACMECertificate dns1 -Generate -AlternativeIdentifierRefs dns2,dns3,dns4 -Alias multiNameCert`

_after_ submitting Complete-ACMEChallenge for one dns?

**Comment by bseddon** [comment]
Yes.  But you will not be successful issuing the New-ACMECertificate command until you have issued the Complete-ACMEChallenge and Submit-ACMEChallenge and until the Update_ACMEIdentifier returns a valid challenge status for each alias dns1, dns2, dns3 and dns4.

**Comment by ebekker** [comment]
I'll use this ticket to track adding support for the cert revocation feature.
